packages/compiler/src/core/program.ts

```diff
@@ -151,7 +157,35 @@ export async function compile(
   mainFile: string,
   options: CompilerOptions = {},
   oldProgram?: Program, // NOTE: deliberately separate from options to avoid memory leak by chaining all old programs together.
-): Promise<Program> {
+) {
+  const logger = createLogger({ sink: host.logSink });
+
+  let program, shouldEmit;
+  if (logger.trackAction) {
+    ({ program, shouldEmit } = await logger.trackAction(
+      () => runCompiler(host, mainFile, options, oldProgram),
+      { level: "trace", message: "Compiling..." },
+    ));
+  } else {
+    ({ program, shouldEmit } = await runCompiler(host, mainFile, options, oldProgram));
+  }
+
+  // Emitter stage
+  if (shouldEmit) {
+    for (const emitter of program.emitters) {
+      await emit(emitter, options, program);
+    }
+  }
+
+  return program;
+}
+
+export async function runCompiler(
+  host: CompilerHost,
+  mainFile: string,
+  options: CompilerOptions = {},
+  oldProgram?: Program,
+): Promise<{ program: Program; shouldEmit: boolean }> {
```

Comment:
```suggestion
): Promise<{ program: Program; shouldAbort: boolean }> {
```

I think this would be less confusing